coins/ethereum/contracts/Router.sol

```diff
@@ -0,0 +1,101 @@
+//SPDX-License-Identifier: AGPLv3
+pragma solidity ^0.8.0;
+
+import "./Schnorr.sol";
+
+contract Router is Schnorr {
+    // contract owner
+    address owner;
+
+    // nonce is incremented for each batch of transactions executed
+    uint256 nonce; 
+
+    // prevents re-entrancy
+    uint8 internal locked;
+
+    struct PublicKey {
+        uint8 parity;
+        bytes32 px;
+    }
+
+    // current aggregated validator public key 
+    PublicKey publicKey;
+
+    struct Transaction {
+        address to;
+        uint256 value;
+        uint256 gas;
+        bytes data;
+    }
+
+    struct Signature {
+        bytes32 e;
+        bytes32 s;
+    }
+
+    event Executed(uint256 nonce, uint256 index, bool success);
+
+    constructor() {
+        owner = msg.sender;
+    }
+
+    modifier onlyOwner() {
+        require(msg.sender == owner, "must be called by the contract owner");
+        _;
+    }
+
+    modifier noReentrant() {
+        require(locked == 0, "no re-entrancy");
+        locked = 1;
+        _;
+        locked = 0;
+    }
+
+    function getNonce() external view returns (uint256) {
```

Comment:
\- for public

Reply:
why public vs external? it's not called within this contract

Reply:
We can remove this function since nonce is available via nonce, since the variable is public. That was my comment.

Reply:
done